### PR TITLE
Optimize `box(x) == null` to `false` and `box(x) != null` to `true`

### PR DESCRIPTION
**Pull request by EgorBo**
Partially addresses https://github.com/mono/mono/issues/17858 (2nd case)
E.g.:
```csharp
static bool Generic<T>(T a)
{
    return a == null;
}


static bool Test()
{
    return Generic(42);
}
```
Codegen for `Test` (was):
```asm
   0:   50                      push   %rax
   1:   48 b8 e0 27 a7 01 00    movabs $0x1a727e0,%rax
   8:   00 00 00
   b:   bf e0 80 92 01          mov    $0x19280e0,%edi
  10:   be 14 00 00 00          mov    $0x14,%esi
  15:   ff 10                   callq  *(%rax)
  17:   c7 40 10 2a 00 00 00    movl   $0x2a,0x10(%rax)
  1e:   31 c0                   xor    %eax,%eax
  20:   59                      pop    %rcx
  21:   c3                      retq
```

Codegen for `Test` (now):
```asm
   0:   31 c0                   xor    %eax,%eax
   2:   c3                      retq
```
The only thing I am not sure about is CoreCLR checks for side-effects for BOX operation for the exactly the same optimization: https://github.com/dotnet/coreclr/blob/a9f3fc16483eecfc47fb79c362811d870be02249/src/jit/gentree.cpp#L12641-L12643 any ideas of such side-effects?

### CODE REVIEW

File: mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -8938,6 +8938,33 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
...
+					is_eq = TRUE;
+				else if (ip = il_read_op (ldnull_ip, end, CEE_PREFIX1, MONO_CEE_CGT_UN))
+					is_neq = TRUE;
+
+				if ((is_eq || is_neq) && ip_in_bb (cfg, cfg->cbb, ip)) {
+					next_ip = ip;
+					il_op = is_eq ? CEE_LDC_I4_0 : CEE_LDC_I4_1;
+					EMIT_NEW_ICONST (cfg, ins, is_eq ? 0 : 1);
+					ins->type = STACK_I4;
+					*sp++ = mono_decompose_opcode (cfg, ins);
```
Comment (vargaz) [inline mono/mini/method-to-ir.c]: I think EMIT_NEW_ICONST already sets this.

Comment (vargaz) [reply mono/mini/method-to-ir.c]: ICONST doesn't need to be decomposed.

**Comment by filipnavara** [comment]
> any ideas of such side-effects?

It's an assert, so I don't think there are any possible side-effect. They just want to enforce that invariant.

The way you implemented the check is actually a stricter version of what CoreCLR does (which is ok) so it avoids any possibility for `call` sneaking in and breaking the optimization.

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build Linux i386

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by filipnavara** [comment]
The Linux i386 failure is interesting:

```
    System.Linq.Parallel.Tests.MinTests.Min_Float(labeled: Default, count: 1, min: 0) [FAIL]
      Assert.Equal() Failure
      Expected: NaN
      Actual:   0
      Stack Trace:
          at System.Linq.Parallel.Tests.MinTests.Min_Float (System.Linq.Parallel.Tests.Labeled`1[T] labeled, System.Int32 count, System.Single min) [0x00082] in /mnt/jenkins/workspace/test-mono-pull-request-i386/external/corefx/src/System.Linq.Parallel/tests/QueryOperators/MinTests.cs:121
```

It corresponds to this line:
https://github.com/mono/corefx/blob/master/src/System.Linq.Parallel/tests/QueryOperators/MinTests.cs#L121

so based on the output the "expected" value should be 0 but it is not. Am I reading it wrong?

File: mono/mini/method-to-ir.c [hunk 2]
```diff
@@ -8938,6 +8938,7 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				}
 			}
 
+#ifdef ENABLE_NETCORE
```
Comment (Therzok) [inline mono/mini/method-to-ir.c]: _why_ do we keep disabling these optimizations for desktop?

Comment (Therzok) [reply mono/mini/method-to-ir.c]: cc @slluis

Comment (Therzok) [reply mono/mini/method-to-ir.c]: cc @marek-safar

Comment (marek-safar) [reply mono/mini/method-to-ir.c]: This one does not need to be ENABLE_NETCORE

Comment (Therzok) [reply mono/mini/method-to-ir.c]: There are quite a few other PRs than went in over the past months which have been enabled in .netcore only and I think they shouldn't be.

I can compile a list, they would be nice to have for vsmac too

Comment (EgorBo) [reply mono/mini/method-to-ir.c]: Ah ok, will enable it back. I thought we only do such changes for netcore only.

Comment (vargaz) [reply mono/mini/method-to-ir.c]: Having them ifdef NETCORE reduces the risk of regression in the production version.

**Comment by lewing** [comment]
I'm going to land this with the check but I'd also like to see some of these optimizations enabled outside of netcore (especially for new platforms)

**Pull request merged**